Add compiler options and metadata reference information into portable (#44373)
dotnet/roslyn: docs/features/compilation-from-portable-pdb.md

This document is restricted to the following assumptions:

1. The benefit is for builds with `-deterministic` and published to the symbol server.
2. Source generator and analyzer references are not needed for this task. They may be useful, but are out of scope for this document.
3. Any storage capacity used for PDBs and source should not impact this feature, such as compression algorithm.
4. Only Portable PDB files will be included for this spec. We can expand this feature past these once it is implemented and proven needed elsewhere.

This document will provide the expanded specification to the Portable PDB format. Any additions to that format will be ported to expand documentation provided in [dotnet-runtime](https://github.com/jnm2/dotnet-runtime/blob/26efe3467741fe2a85780b2d2cd18875af6ebd98/docs/design/specs/PortablePdb-Metadata.md#source-link-c-and-vb-compilers).

## List of Compiler Flags

#### CSharp Flags Not Included

* appconfig
* baseaddress
* bugreport
* checksumalgorithm 
    - Already stored in PDB for each source file

> * debug

Reviewer: Think we should differentiate between flags which are not encoded at all vs. the ones that aren't encoded here but can be re-inferred from other sources.

Reviewer: Basically we should break these up into different sections. Should be easy at a glance to see what arguments are not re-constructable after the fact.

Author: Just make a different section entirely?

Author: Got it, I'll make different sections

## compilation-from-portable-pdb

> Compilation from portable PDBs today is not completely possible, but is desirable in order to help reconstruct a compilation from source provided via source link or embedded in a pdb. Motivation is derived from [roslyn 41395](https://github.com/dotnet/roslyn/issues/41395).

Reviewer 2: @ryzngard Could you spell out an end-to-end scenario? In particular, if this information is embedded in a PDB, how would one get the information out? That should probably be useful to document here.

Reviewer 3: Agreed, we should mention the information is available in CustomDebugInformation table under specific GUIDs. 
Once we finalize the design I also plan to update the [Portable PDB spec](https://github.com/dotnet/runtime/blob/master/docs/design/specs/PortablePdb-Metadata.md#CustomDebugInformationTable) to list these new CDIs. We should also update mdv (MetadataVisualizer) and Pdb2Xml to recognize these.

Author: Added examples of how to get information out of the pdb. Still writing more information for end-to-end scenario

Author: #44703 tracks end-to-end integration test for us. I added a note in the markdown file that scenario should better be explained once that's done, next to a sentence with the overarching goal

This document is restricted to the following assumptions:

1. The benefit is for builds with `-deterministic` and published to the symbol server.
2. Source generator and analyzer references are not needed for this task. They may be useful, but are out of scope for this document.
3. Any storage capacity used for PDBs and source should not impact this feature, such as compression algorithm.
4. Only Portable PDB files will be included for this spec. We can expand this feature past these once it is implemented and proven needed elsewhere.

This document will provide the expanded specification to the Portable PDB format. Any additions to that format will be ported to expand documentation provided in [dotnet-runtime](https://github.com/jnm2/dotnet-runtime/blob/26efe3467741fe2a85780b2d2cd18875af6ebd98/docs/design/specs/PortablePdb-Metadata.md#source-link-c-and-vb-compilers).

## List of Compiler Flags

> #### CSharp Flags That Can Be Derived From PDB

Reviewer 3: >PDB [](start = 43, length = 3)

... or the assembly

You can add the following:
- `baseaddress`, `filealign`, `highentropyva`, `platform`, `publicsign`, `subsystemversion`, `target` in PE headers
- `debug`: PDB format (`portable`, `embedded`, `full`) is stored in PE debug directory
- `debug+`: Included in OptimizationLevel
- `deterministic`: Reproducible record in PE debug directory
- `embed`: Embedded source files included in PDB
- `linkresource`, `resource`: Resource metadata
- `nowin32manifest`, `win32icon`, `win32manifest`, `win32res`: Native resource data in PE

* checksumalgorithm
* link
    - Represented by a metadata reference with `EmbededInteropTypes=true` 
* main
    - Already stored in PDB as the entry point token
* moduleassemblyname
* modulename

#### CSharp Flags Not Included

* appconfig
* baseaddress
* bugreport
* debug
* delaysign
* doc
* deterministic
* embed
* errorendlocation
* errorlog
* errorreport
* filealign
* fullpaths
* highentropyva
* incremental
* keycontainer
* keyfile
* linkresource
* noconfig
* nologo

> * nostdlib

Reviewer 3: >nostdlib [](start = 2, length = 8)

`nostdlib` is reflected in the new *Metadata References* custom debug info.

* nowarn
* nowin32manifest

> * optimize

Reviewer 3: >optimize [](start = 2, length = 8)

`optimize` - we should include OptimizationLevel in the options we write to the PDB.

* checksumalgorithm
* main 

#### Visual Basic Flags Not Included

* baseaddress
* bugreport
* debug
* delaysign
* doc
* errorreport
* filealign
* help
* highentropyva
* keycontainer
* keyfile
* libpath
* linkresource
* moduleassemblyname
* modulename
* netcf
* noconfig
* nologo
* nostdlib
* nowarn
* nowin32manifest
* optimize
* optioncompare
* optionexplicit
* optioninfer
* out
* parallel
* platform
* preferreduilang
* quiet
* recurse
* refonly
* refout
* resource
* rootnamespace
* ruleset
* sdkpath
* subsystemversion
* target
* utf8output
* vbruntime
* verbose
* warnaserror
* win32icon
* win32manifest
* win32resource

#### Compiler Version

This will be a full compiler version, including commit SHA for the build of the compiler. This will be used since compilations across compiler versions are not guaranteed to be the same.

#### Source Encoding

Encoding for source files that are not declared using BOM. This 

#### Metadata References

Symbol server uses a [key](https://github.com/dotnet/symstore/blob/master/docs/specs/SSQP_Key_Conventions.md#pe-timestamp-filesize) computed from the COFF header in the PE image:

Timestamp: 4 byte integer
Size of image: 4 byte integer

Example:

> File name: `Foo.exe` 

Reviewer 2: >Foo [](start = 12, length = 3)

nit: policheck opposes usage of "foo"

Example: 

`compilerversion\01.2.3\0sourceencoding\0utf-8\0checked\01\0unsafe\01\0langversion\0latest\0nullable\0Enable` 

##### Options For CSharp

See [compiler options](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/listed-alphabetically) documentation

1. "compilerversion"
2. "[codepage](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/codepage-compiler-option)"
3. "[checked](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/checked-compiler-option)"
4. "[unsafe](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/unsafe-compiler-option)"
5. "[langversion](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/langversion-compiler-option)"
6. "[nullable](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references)"

> 7. "[define](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/define-compiler-option)"

Reviewer 3: >define [](start = 5, length = 6)

Missing : `optimizations` (CompilationOptions.OptimizationLevel)

##### Options For Visual Basic

See [compiler options](https://docs.microsoft.com/en-us/dotnet/visual-basic/reference/command-line-compiler/compiler-options-listed-alphabetically) documentation

> 1. "compilerversion"

Reviewer 3: >compilerversion [](start = 4, length = 15)

We also need to add `runtime-version`, which would be `(AssemblyInformationalVersionAttribute)typeof(object).Assembly.GetCustomAttributes(typeof(AssemblyInformationalVersionAttribute), false).Single().InformationalVersion`

On desktop .NET the version is e.g. `4.8.4180.0`, on .NET Core the version is e.g. `3.1.2-servicing.20066.2+fd22fd600a2526a7c58536da0e80325eb710d7f1`.

Reviewer: Why do we need the runtime version in the PDB? Yeah there are some really odd corner cases where this would be relevant but in general it should not be. Seems odd to include this but not for instance the underlying operating system (Linux vs. Windows) which is much more likely to impact the deterministic output of the compilation.

Reviewer 3: The runtime version affects Unicode character interpretation and decimal arithmetics. What differences are there between Linux and Windows? We expects paths to be normalized thru path map.

Reviewer 3: The only OS-level differences I am aware of are in deflate compression implementation, since CLR currently calls into the OS to implement compression. This affects PDB embedded source files and embedded PDBs. Ideally, we would include the version of the native compression library as well. For that we will need API from CLR to retrieve the version. Alternatively, we can compare the decompressed versions of these artifacts for equality. But then we would also need to ignore PDB checksum when comparing the binaries.

Reviewer 3: The tool that does the validation can download the specific version of CLR (or a container with that version of the CLR) and run in it. The same with compression lib, but it's harder to do. We could include the full OS version, but getting exact versions of various Linux distros is gonna be problematic (I think). Limiting the problem to finding a specific version of compression library seems easier.

Reviewer: > What differences are there between Linux and Windows? 

The different line endings impacts determinism because it changes the checksum of source files across Linux / Windows. Even if you force consistent line endings in your source code across OS via .gitattributes (most repos I've seen don't do this) you still have to deal with the generated sources that get included in your builds. Last time I looked at this I saw generators using OS specific line endings which meant they would be different across Linux / Windows.

Reviewer 3: The plan for line endings to record git options in Source Link: https://github.com/dotnet/sourcelink/issues/613. This should allow us to clone the repo with exactly the same line breaks that the compiler saw on the build server.

Reviewer 3: I guess one disadvantage of including the runtime version is that now PDB/DLL actually depends on the runtime version even in cases when they did not before (similar problem like the encoding). We could potentially make this more fine-grained if the CLR exposed version of Unicode it supports and version of decimal operation implementation (not sure if that's actually viable). @richlander

Reviewer 3: Thinking about this more - the usual case is that the compiler comes with the runtime in an SDK (on .NET Core), right? It is likely then that whenever compiler version changes the runtime changes as well.

Reviewer: > The plan for line endings to record git options in Source Link: dotnet/sourcelink#613. This should allow us to clone the repo with exactly the same line breaks that the compiler saw on the build server

How will this fix generated files though? Think about generated assembly info. That comes from an MSBuild task which eventually (through `CodeDomProvider` and `TextWriter`) uses `Environment.NewLine` in the `AssemblyInfo.cs` file. That in turn will get included in the PDB file and seemingly (at least that I can think through) will have different line endings on Linux / Windows. 

Did want to ask, but your message seems to confirm, whether or not actually building repos was a goal. Seems like it is though.

Reviewer 3: The files that are not checked into the repo are going to be embedded in the PDB. When reconstructing the compilation we are not gonna run the code generator or any msbuild targets. We are gonna take the files from the PDB and add them directly to the Compilation.

Reviewer 3: > whether or not actually building repos was a goal.

It's not a goal to run msbuild. We are gonna bypass all msbuild and just create Compilation and call Emit on it to get the binary.

4. Only Portable PDB files will be included for this spec. We can expand this feature past these once it is implemented and proven needed elsewhere.

This document will provide the expanded specification to the Portable PDB format. Any additions to that format will be ported to expand documentation provided in [dotnet-runtime](https://github.com/jnm2/dotnet-runtime/blob/26efe3467741fe2a85780b2d2cd18875af6ebd98/docs/design/specs/PortablePdb-Metadata.md#source-link-c-and-vb-compilers).

## List of Compiler Flags

#### CSharp Flags That Can Be Derived From PDB or Assembly

* baseaddress
* checksumalgorithm
* debug
* deterministic
* embed
* filealign
* highentropyva
* link
    - Represented by a metadata reference with `EmbededInteropTypes=true` 
* linkresource
    - Will be represented in metadata reference embedded in pdb
* main
    - Already stored in PDB as the entry point token
* moduleassemblyname
* modulename
* nowin32manifest
* pdb
* platform
* publicsign
* resource
    - Will be represented in metadata reference embedded in pdb
* subsystemversion
* target
* win32icon
* win32manifest
* win32res

#### CSharp Flags Not Included

> * appconfig

Reviewer: This impacts assembly identity resolution. If this is omitted then it's possible there won't be enough info in the PDB to successfully re-creeate compilations.

Author: Is the appconfig file included in embedded resources or sourcelink somewhere? If we include this, it may not be possible to get the same file again

Reviewer: No it's not included. It's simply consumed at compilation time. But it is necessary to reproduce the compilation given the same inputs (assuming of course the appconfig was provided because it was needed to resolve reference dependency issues)

Reviewer 3: Yes, you're right. I forgot about this one. It's needed in order to create `AssemblyIdentityComparer`. The only data needed from the config file are stored in [`AssemblyPortabilityPolicy`](http://sourceroslyn.io/#Microsoft.CodeAnalysis/Desktop/AssemblyPortabilityPolicy.cs,14). There are two bools we need to serialize. We can, for example, serialize this as option `portability-policy` = number in [0..3] (1 bit for each bool).

* bugreport
* delaysign
* doc
* errorendlocation
* errorlog
* errorreport
* fullpaths
* incremental
* keycontainer

> * keyfile

Reviewer: All of the key file options meaningfully impact compilation output: delaysign, keyfile and keycontainer. Why are they not being included?

Author: If something is signed we can't store the private keys anywhere, and public keys are already available. Maybe worth calling out in this documentation, but I don't know how we would use this information

Reviewer: One of the goals here is to be able to accurately re-produce PE files. Given that our recommendation to library authors is still to strong name sign their PE files not having this seems like a problematic gap.

Reviewer 3: We can 100% reproduce strong name - the public key is stored in AssemblyDef metadata table.
Reproducing signature without private key is impossible. The resulting binary will differ in authenticode signature and strong name signature (they will be empty). You can however sign the binary with the private key if you have it and the result will be exactly the same binary.

Reviewer 3: We definitely need tests for various signing cases.

Reviewer 3: I think the tool that re-creates the compilation for validation can set CompilationOptions.PublicSign = true if the COR flags in the DLL have StrongNameSignature bit set, otherwise it would use no signing. That should produce the same binary modulo the actual signatures, correct?

Reviewer: > That should produce the same binary modulo the actual signatures, correct?

That matches my understanding.

Reviewer: Thought a bit more about this last night. I think it's okay to deliberately omit the key sign data. The only interesting piece of information is the actual public key which is in the assembly. 

The goal here is to be able to verify, by recreating, that this is the correct binary for the given inputs. When it comes to strong names we can simply take the binary we're verifying and strip out the strong name info. Essentially undo strong name and put the binary back into a `publicSign` state. At that point the information in the PDB should be enough to recreate a matching binary (byte for byte)

Reviewer 3: Yup, exactly.

* noconfig
* nologo

> * nostdlib

Reviewer: This is an important option cause it impacts compilation. Can it be derived from the existing PDB info?

Author: nostdlib can be found because of metadata references, but I'm not sure how much it correlates directly. I meant to move this up and might have missed it

Author: I put in VB but not C# portion. Just needs to be moved

Reviewer 3: `nostdlib` is only [used](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs#L1324) by the command line compiler when deciding what references to create for the compilation. We store a complete set of metadata references that were passed to the compilation in the PDB, so we don't need `nostdlib`.

Reviewer: That's true. Essentially when re-creating the command line, should we choose to do that, we should always pass `/nostdlib` because what we store in the PDB is the final reference set.

Reviewer 3: Yes. I believe it will be possible to create a command line, it won't obviously be the exactly same command line, but it should produce the same results (modulo signing and 
perhaps the PDB compression depending how deal with the compression lib). 

Note that the main goal is not really the command line, but rather a Compilation with all the options and parameters necessary. But as I said text version of the command line should also be possible.

* nowarn
* out
* parallel
* pathmap
* preferreduilang
* recurse
* refout
* refonly
* reportanalyzer

> * resource

Reviewer: Values passed to this option are embedded in the produced PE file and hence impacts the output.

Author: are embedded resources included in the PE header?

Reviewer: Embedded resources are included in the PE file, not exactly sure where. @tmat will know.

Reviewer 3: Indeed they are, therefore they can be extracted from the PE file and we do not need to store them in the PDB.

* reportanalyzer
* ruleset
* utf8output
* version
* warn
* warnaserror

#### Visual Basic Flags That Can Be Derived From PDB or Assembly

* baseaddress
* checksumalgorithm
* debug
* filealign
* highentropyva
* linkresource
    - Will be represented in metadata reference embedded in pdb
* main 
* nostdlib
    - Will be represented in metadata reference embedded in pdb
* platform
* resource
    - Will be represented in metadata reference embedded in pdb
* subsystemversion
* target
* win32icon
* win32manifest
* win32resource

#### Visual Basic Flags Not Included

* bugreport
* delaysign
* doc
* errorreport
* help
* keycontainer
* keyfile
* libpath
* moduleassemblyname
* modulename
* netcf
* noconfig
* nologo
* nowarn
* nowin32manifest
* optimize
* optioncompare
* optionexplicit
* optioninfer
* out
* parallel
* preferreduilang
* quiet
* recurse
* refonly
* refout
* rootnamespace
* ruleset
* sdkpath
* utf8output
* vbruntime
* verbose
* warnaserror

#### Compiler Version

This will be a full compiler version, including commit SHA for the build of the compiler. This will be used since compilations across compiler versions are not guaranteed to be the same.

#### Source Encoding

> Encoding for source files that are not declared using BOM. This 

Reviewer: Need to finish out this paragraph. Presently ends with an incomplete sentence.

Author: woops 😑 adding more info about how this is retrieved and stored

#### Metadata References

Symbol server uses a [key](https://github.com/dotnet/symstore/blob/master/docs/specs/SSQP_Key_Conventions.md#pe-timestamp-filesize) computed from the COFF header in the PE image:

Timestamp: 4 byte integer
Size of image: 4 byte integer

Example:

File name: `example.exe` 

COFF header Timestamp field: `0x542d5742` 

COFF header SizeOfImage field: `0x32000` 

Lookup key: `example.exe/542d574232000/example.exe` 

To fully support metadata references, we'll need to be able to find the exact PE image that was used in the compilation. We'll do this by storing the parts that make up the symbol server key.

We will also store the MVID of a reference since it's a GUID that represents the symbol. This is to future proof the information we will need to look up references.

At this time, we'll only include external references for the compilation. Any other references may be added later in a separate blob.

## PDB Format Additions

#### Metadata References

Metadata references can be easily stored as binary. The binary encoding will be as follows (order matters):

Name: A UTF-8 string (null terminated)

> Aliases: UTF-8 Comma (, ) separate list of aliases (null terminated). May be empty

Reviewer 3: "separated"

MetadataImageKind: byte value representing Microsoft. CodeAnalysis. MetadataImageKind

EmbedInteropTypes/MetadataImageKind: byte, reads as binary values starting in order from right to left

>     

Reviewer: Double blank line

Author: Needed for formatting

This is what the preview looks like in vscode

![image](https://user-images.githubusercontent.com/475144/83439887-cfb23100-a3f8-11ea-913c-2499b1db85ca.png)

Author: FWIW you can use the rich diff in GitHub to see diff of what is generated. I just moved a bunch of stuff around to hopefully help with readability and reduce redundant statements

            Console.WriteLine(name);
            Console.WriteLine($"Extern Aliases: \"{externAliases}\"");
            Console.WriteLine($"Embed Interop Types: {embedInteropTypes}");
            Console.WriteLine($"Metadata Image Kind: {kind}");
            Console.WriteLine($"Timestamp: {timestamp}");
            Console.WriteLine($"Image Size: {imageSize}");
            Console.WriteLine($"MVID: {mvid}");
            Console.WriteLine();
        }
    }
}
```

#### Compiler Flag Key Value Pairs

The remaining values will be stored as key value pairs in the pdb. The storage format will be UTF8 encoded key value pairs that are null terminated. Order is not guaranteed. Any values left out can be assumed to be the default for the type. Keys may be different for Visual Basic and CSharp. They are serialized to reflect the command line arguments representing the same values

Example: 

`compilerversion\01.2.3-example-sha\0sourceencoding\0utf-8\0checked\01\0unsafe\01\0langversion\0latest\0nullable\0Enable` 

##### Options For CSharp

See [compiler options](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/listed-alphabetically) documentation

> * "[appconfig](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/appconfig-compiler-option)"

Reviewer 3: >appconfig [](start = 4, length = 9)

This is somewhat confusing - the key in the PDB is called `portability-policy` but here you mention `appconfig`, which is the name of the command line switch. On the other hand, "compilerversion" below is the PDB key. I think it would be good to clearly differentiate between these.
Perhaps make this a table with first column being the PDB key, the second corresponding command line switch and third description.

Author: Made a table, think it is much clearer. Linked complex topics to other sections in the document.

        - 0 -> SuppressSilverlightPlatformAssembliesPortability = false, SuppressSilverlightLibraryAssembliesPortability = false
        - 1 -> SuppressSilverlightPlatformAssembliesPortability = true, SuppressSilverlightLibraryAssembliesPortability = false
        - 2 -> SuppressSilverlightPlatformAssembliesPortability = false, SuppressSilverlightLibraryAssembliesPortability = true
        - 3 -> SuppressSilverlightPlatformAssembliesPortability = true, SuppressSilverlightLibraryAssembliesPortability = true
* "[checked](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/checked-compiler-option)"
* "[codepage](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/codepage-compiler-option)"

> * "compilerversion"

Author: I thought we discussed that runtime version wouldn't be needed. I'm going to reread https://github.com/dotnet/roslyn/pull/44373#discussion_r431280947

> * "compilerversion"

Author: I don't have a strong opinion on multiple words with hyphen, I just did it as is because that's how cli works. Happy to go either way if you feel strongly about it.

* "[define](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/define-compiler-option)"
* "[langversion](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/langversion-compiler-option)"
* "[nullable](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references)"

> * "[optimize](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/optimize-compiler-option)"

Reviewer 3: >optimize [](start = 4, length = 8)

How is `optimize` encoded?

Author: It's a short note at the beginning that they are all UTF-8 strings, serialized to reflect commandline values. optimize+ for example would be a boolean value serialized as UTF-8 string

* main 
* nostdlib
    - Will be represented in metadata reference embedded in pdb
* platform
* resource
    - Will be represented in metadata reference embedded in pdb
* subsystemversion
* target
* win32icon
* win32manifest
* win32resource

#### Visual Basic Flags Not Included

* bugreport
* delaysign
* doc
* errorreport
* help
* keycontainer
* keyfile
* libpath
* moduleassemblyname
* modulename
* netcf
* noconfig
* nologo
* nowarn
* nowin32manifest
* optimize
* optioncompare
* optionexplicit
* optioninfer
* out
* parallel
* preferreduilang
* quiet
* recurse
* refonly
* refout
* rootnamespace
* ruleset
* sdkpath
* utf8output
* vbruntime
* verbose
* warnaserror

#### Options For CSharp

See [compiler options](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/listed-alphabetically) documentation

| PDB Key | Documentation   |
| ------- | --------------- |
| portability-policy | see [portability policy](#portability-policy) |
| checked | [checked](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/checked-compiler-option) |
| default-encoding | see [file encoding](#file-encoding) |

> | compilerversion | full version with SHA |

Reviewer 3: >compilerversion [](start = 2, length = 15)

Missing `-`.

Author: I'm just going to standardize on using `-` everywhere. I think I like it more, since it's more readable to humans.

    - 0 -> SuppressSilverlightPlatformAssembliesPortability = false, SuppressSilverlightLibraryAssembliesPortability = false
    - 1 -> SuppressSilverlightPlatformAssembliesPortability = true, SuppressSilverlightLibraryAssembliesPortability = false
    - 2 -> SuppressSilverlightPlatformAssembliesPortability = false, SuppressSilverlightLibraryAssembliesPortability = true
    - 3 -> SuppressSilverlightPlatformAssembliesPortability = true, SuppressSilverlightLibraryAssembliesPortability = true

#### File Encoding 

Encoding will be stored in two keys: "default-encoding" and "fallback-encoding". 

If "default-encoding" is present, it represents the forced encoding used, such as by passing in "codepage" to the command line. All files should be decoded using this codepage value.

If "fallback-encoding" is present, it means that no default encoding was specified but an encoding was detected and used. The compiler [has logic](https://github.com/dotnet/roslyn/blob/462eac607741023e5c2d518ac1045f4c6dabd501/src/Compilers/Core/Portable/EncodedStringText.cs#L32) to determine an encoding is none is specified, so the value that was computed is stored so it can be reused in future compilations.

> Both values are codepage values for an encoding, such as `65001` for UTF-8. 

Reviewer 3: >`65001` for UTF-8.  [](start = 57, length = 19)

I think it would be better to use [`Encoding.WebName`](https://docs.microsoft.com/en-us/dotnet/api/system.text.encoding.webname?view=netcore-3.1) because that's more standardized than code pages (code pages are Microsoft specific). Also document it's case-insensitive name.
See also [http://www.iana.org/assignments/character-sets/character-sets.xhtml](http://www.iana.org/assignments/character-sets/character-sets.xhtml)

Author: Done, I didn't know CodePage was Microsoft specific

* optimize
* optioncompare
* optionexplicit
* optioninfer
* out
* parallel
* preferreduilang
* quiet
* recurse
* refonly
* refout
* rootnamespace
* ruleset
* sdkpath
* utf8output
* vbruntime
* verbose
* warnaserror

#### Options For CSharp

See [compiler options](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/listed-alphabetically) documentation

| PDB Key | Documentation   |
| ------- | --------------- |
| portability-policy | see [portability policy](#portability-policy) |
| checked | [checked](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/checked-compiler-option) |
| default-encoding | see [file encoding](#file-encoding) |
| compiler-version | full version with SHA |
| fallback-encoding | see see [file encoding](#file-encoding) |
| define | [define](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/define-compiler-option) |
| language-version | [langversion](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/langversion-compiler-option) |
| nullable | [nullable](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references) |

> | optimize | see [optimization](#optimization) |

Reviewer 3: >optimize [](start = 2, length = 8)

I think this is now optimization

#### Optimization

Optimization level can be specified from the command line with [optimize](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/optimize-compiler-option). It gets translated to the [OptimizationLevel](https://github.com/dotnet/roslyn/blob/e704ca635bd6de70a0250e34c4567c7a28fa9f6d/src/Compilers/Core/Portable/Compilation/OptimizationLevel.cs) that is emitted to the PDB. That means it will be stored as the enum, not a boolean value like the command line switch.

### Options For Visual Basic

See [compiler options](https://docs.microsoft.com/en-us/dotnet/visual-basic/reference/command-line-compiler/compiler-options-listed-alphabetically) documentation

| PDB Key | Documentation   |
| ------- | --------------- |
| compiler-version | full version with SHA |
| define | [define](https://docs.microsoft.com/en-us/dotnet/visual-basic/reference/command-line-compiler/define) |
| language-version | [langversion](https://docs.microsoft.com/en-us/dotnet/visual-basic/reference/command-line-compiler/langversion) |

> | optimize | see [optimization](#optimization) |

Reviewer 3: >optimize [](start = 2, length = 8)

I think this is now `optimization`

Since appconfig is a pointer to a file not embedded in the PDB or PE, information that will directly impact the compilation is extracted. This is stored in a flag called `portability-policy` with a numeric value from \[0-3\]. This value directly correlates to reconstructing the [AssemblyPortabilityPolicy](https://github.com/dotnet/roslyn/blob/bdb3ece74c85892709f5e42ae7d67248999ecc3b/src/Compilers/Core/Portable/Desktop/AssemblyPortabilityPolicy.cs).

* 0 -> SuppressSilverlightPlatformAssembliesPortability = false, SuppressSilverlightLibraryAssembliesPortability = false
* 1 -> SuppressSilverlightPlatformAssembliesPortability = true, SuppressSilverlightLibraryAssembliesPortability = false
* 2 -> SuppressSilverlightPlatformAssembliesPortability = false, SuppressSilverlightLibraryAssembliesPortability = true
* 3 -> SuppressSilverlightPlatformAssembliesPortability = true, SuppressSilverlightLibraryAssembliesPortability = true

#### File Encoding 

Encoding will be stored in two keys: "default-encoding" and "fallback-encoding". 

> If `default-encoding` is present, it represents the forced encoding used, such as by passing in "codepage" to the command line. All files should be decoded using this encoding.

Author: Yes, modifying to "All files without a Byte Order Mark (BOM) should be decoded using this encoding"

## compilation-from-portable-pdb

Compilation from portable PDBs today is not completely possible, but is desirable in order to help reconstruct a compilation from source provided via source link or embedded in a pdb. Motivation is derived from [roslyn 41395](https://github.com/dotnet/roslyn/issues/41395).

This document is restricted to the following assumptions:

1. The benefit is for builds with `-deterministic` and published to the symbol server.

> 2. Source generator and analyzer references are not needed for this task. They may be useful, but are out of scope for this document.

Reviewer: Will the analyzers and generators used be recorded? I 100% agree we don't need their actual references to form a functional compilation. Yet if we're recording information about compilation to create the source we should record that a given analyzer or generator was used. That can help for instance us route bugs to the appropriate generator for bad code or to an analyzer saying "hey you missed a case you were trying to prevent"

Reviewer 3: We can include them in the set of metadata references if you think it's valuable. Note that we can't reconstruct a compilation that we could use to re-run the analyzers exactly as they were executed during build without also including additional files and perhaps other information the analyzers/generators need from build.

Reviewer 4: Almost impossible, as we would have to find their transitive references as well, which is dangerous.

Reviewer: I'm more concerned with just knowing which analyzer and SG run, not their transitive references. The assembly qualified name is essentially what I'm imagining here.

Reviewer 3: I suggest we consider including analyzer/generator assembly refs separately from this change. The goal of this change is to enable reconstructing compilation specifically for the purpose of validation that a DLL was compiled from given sources and for running additional analyzers after build. The format of the PDB info can be expanded in future with more information to support more scenarios.

3. Any storage capacity used for PDBs and source should not impact this feature, such as compression algorithm.

> 4. Only Portable PDB files will be included for this spec. We can expand this feature past these once it is implemented and proven needed elsewhere.

Reviewer: This will break round tripping between native and portable PDBs. Don't our own tools depend on this being possible?

Reviewer 3: We don't depend on it. The tests that compare PDB formats already account for other differences between the formats.

Author: Marking as resolved so I can keep track of open concerns. Feel free to reopen if it's not.

This document will provide the expanded specification to the Portable PDB format. Any additions to that format will be ported to expand documentation provided in [dotnet-runtime](https://github.com/jnm2/dotnet-runtime/blob/26efe3467741fe2a85780b2d2cd18875af6ebd98/docs/design/specs/PortablePdb-Metadata.md#source-link-c-and-vb-compilers).

## List of Expanded Information

#### Compiler Version

> This will be a full compiler version, and if possible the commit SHA for the build of the compiler. This will be used since compilations accros compiler versions are not guaranteed to be the same.

Author: ```suggestion
This will be a full compiler version, including commit SHA for the build of the compiler. This will be used since compilations across compiler versions are not guaranteed to be the same.
```

3. "[checked](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/checked-compiler-option)"
4. "[unsafe](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/unsafe-compiler-option)"
5. "[langversion](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/langversion-compiler-option)"
6. "[nullable](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references)"
7. "[define](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/define-compiler-option)"

##### Options For Visual Basic

See [compiler options](https://docs.microsoft.com/en-us/dotnet/visual-basic/reference/command-line-compiler/compiler-options-listed-alphabetically) documentation

1. "compilerversion"
2. "[removeintchecks](https://docs.microsoft.com/en-us/dotnet/visual-basic/reference/command-line-compiler/removeintchecks)"
3. "[langversion](https://docs.microsoft.com/en-us/dotnet/visual-basic/reference/command-line-compiler/langversion)"
4. "[define](https://docs.microsoft.com/en-us/dotnet/visual-basic/reference/command-line-compiler/define)"

> 5. "[optionstrict](https://docs.microsoft.com/en-us/dotnet/visual-basic/reference/command-line-compiler/optionstrict)"

Reviewer 5: >[optionstrict] [](start = 4, length = 14)

What drives the decision whether an option should be included or excluded?

Author: The goal is to include enough information in the PDB so that a consumer could fully reconstruct the original compilation based on the information stored in the PDB, the assembly metadata, symbol and source servers and any private keys used to build the assembly.

Any compiler switches needed to meet that goal should be included

Reviewer: > Any compiler switches needed to meet that goal should be included

I'd pose the question in the other direction: what switches can be excluded because they don't meaningfully impact the compilation? In general all switches impact the compilation, there are very few which do not.

Author: * appconfig
* baseaddress
* bugreport
* checksumalgorithm
* debug
* delaysign
* doc
* deterministic
* embed
* errorendlocation
* errorlog
* errorreport
* filealign
* fullpaths
* highentropyva
* incremental
* keycontainer
* keyfile
* link
* linkresource
* main
* moduleassemblyname
* modulename
* noconfig
* nologo
* nostdlib
* nowarn
* nowin32manifest
* optimize
* out
* parallel
* pathmap
* pdb
* platform
* preferreduilang
* publicsign
* recurse
* refout
* refonly
* reportanalyzer
* resource
* ruleset
* subsystemversion
* target
* utf8output
* version
* warn
* warnaserror
* win32icon
* win32manifest
* win32res

That's the list of all compiler switches (that I could find) that I believe are not needed because they are either implicitly or explicitly available through pdb information already, or are not needed to directly reconstruct the compilation in such a way that we could run analyzers on the input to the compilation and be confident that it represents the code and settings used to create the PE file.

Reviewer: * checksumalgorithm

Seems feasible that analyzers would want to audit this. Consider that security analyzers frequently ban older check sum algorithms. 

* debug
* optimize

Where is this already stored that we could reconstruct it?

* link

This is essentially equivalent to `/reference` why would we exclude it?

* main

This meaningfully impacts the compilation, think it needs to be included. 

* moduleassemblyname
* modulename

There are lots of analyzers out there which focus on naming. Why would we exclude any name based arguments?

* nostdlib

This meaningfully impacts compilation. 

Overall I think we need to keep the full list of options in this document including the list of options we explicitly decided to not include. That way future developers have a baseline to consider. Lacking the full set of options in this document a developer can't tell for a given option if we considered and rejected or just didn't consider an option at all.

Reviewer 3: - `checksumalgorithm` is already stored in the PDB for each source file. 
- optimization level should be included
- `link` is already included since it's represented by a metadata reference with EmbedInteropTypes=true. 
- `main` is already included in the PDB (the entry point token). 
- `moduleassemblyname`, `modulename` are in the metadata
- `nostdlib` is not needed as we have encode exact set of all references.

> Overall I think we need to keep the full list of options in this document including the list of options we explicitly decided to not include. That way future developers have a baseline to consider. Lacking the full set of options in this document a developer can't tell for a given option if we considered and rejected or just didn't consider an option at all.

+1

Author: I'll change documentation to include this full list, maybe with notes about why they aren't included

Author: Updated for VBC and CSC, let me know if there needs to be more information. Most of it is just the list of things missing, with a few noted that they already have information in the pdb

Reviewer 6: Question: why not include all switches always?

Few reasons:

1. Easier to reason about. The compilers always take all the switches and store them, no need to parse, include/exclude specific ones. Less room for an error to slip by when introducing a new switch
2. These are baked into assets that will be around for a long time. We don't necessarily know all the possible ways that the values might be useful in the future. Having them in now future-proofs us.

What's the benefit of including/excluding specific switches?

Reviewer: > What's the benefit of including/excluding specific switches?

One concrete reason is that it could break determinism. Consider that some switches in their raw form, such as source file paths, are mutated by other arguments to the compiler like `/pathMap`. Hence if we included all switches literally then it would break determinism because we'd have original paths vs. mapped paths in the PDB. 

So there does need to be some level of processing and exclusion here. Concretely including `pathMap` would in many ways defeat the purpose of having it 😦 

There are others which are questionable but it's hard to reason about because the goals of this are not clear. Is the goal that we need enough information such that we could recreate the binary with byte for byte equivalence? Essentially do we want enough info to guarantee that we can do the following:

1. Enough options are stored such that given the same inputs we could produce the same assembly (byte for byte equivalence)?
1. Enough option information is stored that we could detect when inputs are not the same. Example is that storing the MVID of a DLL is a strong barrier that we are presented with the right input when re-creating the compilation.

Reviewer 3: > Is the goal that we need enough information such that we could recreate the binary with byte for byte equivalence? 

Yes.

> Enough options are stored such that given the same inputs we could produce the same assembly (byte for byte equivalence)?

Yes.

> Enough option information is stored that we could detect when inputs are not the same. Example is that storing the MVID of a DLL is a strong barrier that we are presented with the right input when re-creating the compilation.

We need information to find the metadata references. We include enough information so that we can look them up on symbol server, e.g. (or a similar storage).

Reviewer 3: > Question: why not include all switches always?

That doesn't work for multiple reasons. One is determinism that Jared mentioned. Others are that some info can't be represented in the PDB in the same way it's represented on command line. Examples include references, for which we need ids that we can use to look the files up on symbol server. Another one is resources that will need to be retrieved from the DLL as we don't want to duplicate the resources in the PDB. We also don't want to duplicate other information that's easily retrievable from the DLL (such as various PE header flags) since whenever you have the same info stored in multiple places there is a potential for mismatch.

Reviewer 3: Finally, there is also the aspect of Compilation APIs that are public. Some command line switches are not even available to the Compilation APIs because they are only needed by the command line compiler to find DLLs, sources, etc (such as varioys `/lib`, `/nostdlib` etc). It'd be unnecessary to add public APIs to EmitOptions to pass them in as we don't need these to recreate the compilation. We already have the info in other forms.
